### PR TITLE
Allow for resolution of NPM registry paths that match upstream

### PR DESCRIPTION
**Pull request by Racer159**
This PR fixes issue #21567 allowing for package tarball URLs to match the upstream registry (and GitLab/JFrog Artifactory URLs).  It uses a regex to parse the filename (which contains the NPM version) and does a fuzzy search to pull it out.  The regex was built/expanded from http://json.schemastore.org/package, https://github.com/Masterminds/semver, and https://docs.npmjs.com/cli/v6/using-npm/semver and is testable here: https://regex101.com/r/OydBJq/5

### CODE REVIEW

File: routers/api/packages/api.go [hunk 1]
```diff
@@ -215,6 +215,10 @@ func Routes(ctx gocontext.Context) *web.Route {
 					r.Get("", npm.DownloadPackageFile)
 					r.Delete("/-rev/{revision}", reqPackageAccess(perm.AccessModeWrite), npm.DeletePackageVersion)
 				})
+				r.Group("/-/{filename}", func() {
+					r.Get("", npm.DownloadPackageFile)
+					r.Delete("/-rev/{revision}", reqPackageAccess(perm.AccessModeWrite), npm.DeletePackageVersion)
+				})
```
Comment (KN4CK3R) [inline routers/api/packages/api.go]:
```suggestion
				r.Get("/-/{filename}", npm.DownloadPackageFileByName)
```
The revision endpoints should not be here and they don't work with the changed url.

File: routers/api/packages/api.go [hunk 2]
```diff
@@ -227,6 +231,10 @@ func Routes(ctx gocontext.Context) *web.Route {
...
```
Comment (KN4CK3R) [inline routers/api/packages/api.go]:
```suggestion
				r.Get("/-/{filename}", npm.DownloadPackageFileByName)
```

File: tests/integration/api_packages_npm_test.go [hunk 3]
```diff
@@ -127,10 +127,17 @@ func TestPackageNpm(t *testing.T) {
 		b, _ := base64.StdEncoding.DecodeString(data)
 		assert.Equal(t, b, resp.Body.Bytes())
 
+		req = NewRequest(t, "GET", fmt.Sprintf("%s/-/%s", root, filename))
+		req = addTokenAuthHeader(req, token)
+		resp = MakeRequest(t, req, http.StatusOK)
+
+		b, _ = base64.StdEncoding.DecodeString(data)
```
Comment (KN4CK3R) [inline tests/integration/api_packages_npm_test.go]:
```suggestion
```

**Review by KN4CK3R — COMMENTED** [review]
And `DownloadPackageFileByName` would look like this:
```go
func DownloadPackageFileByName(ctx *context.Context) {
	filename := ctx.Params("filename")

	pvs, _, err := packages_model.SearchVersions(ctx, &packages_model.PackageSearchOptions{
		OwnerID: ctx.Package.Owner.ID,
		Type:    packages_model.TypeNpm,
		Name: packages_model.SearchValue{
			ExactMatch: true,
			Value:      packageNameFromParams(ctx),
		},
		HasFileWithName: filename,
		IsInternal:      util.OptionalBoolFalse,
	})
	if err != nil {
		apiError(ctx, http.StatusInternalServerError, err)
		return
	}
	if len(pvs) != 1 {
		apiError(ctx, http.StatusNotFound, nil)
		return
	}

	s, pf, err := packages_service.GetFileStreamByPackageVersion(
		ctx,
		pvs[0],
		&packages_service.PackageFileInfo{
			Filename: filename,
		},
	)
	if err != nil {
		if err == packages_model.ErrPackageFileNotExist {
			apiError(ctx, http.StatusNotFound, err)
			return
		}
		apiError(ctx, http.StatusInternalServerError, err)
		return
	}
	defer s.Close()

	ctx.ServeContent(pf.Name, s, pf.CreatedUnix.AsLocalTime())
}
```
Then there is no regex parsing necessary. Which fails btw for `pack-age-1.0.0-v1-beta.0.tgz`.

**Comment by Racer159** [comment]
> And `DownloadPackageFileByName` would look like this:
> 
> ```go
> func DownloadPackageFileByName(ctx *context.Context) {
> 	filename := ctx.Params("filename")
> 
> 	pvs, _, err := packages_model.SearchVersions(ctx, &packages_model.PackageSearchOptions{
> 		OwnerID: ctx.Package.Owner.ID,
> 		Type:    packages_model.TypeNpm,
> 		Name: packages_model.SearchValue{
> 			ExactMatch: true,
> 			Value:      packageNameFromParams(ctx),
> 		},
> 		HasFileWithName: filename,
> 		IsInternal:      util.OptionalBoolFalse,
> 	})
> 	if err != nil {
> 		apiError(ctx, http.StatusInternalServerError, err)
> 		return
> 	}
> 	if len(pvs) != 1 {
> 		apiError(ctx, http.StatusNotFound, nil)
> 		return
> 	}
> 
> 	s, pf, err := packages_service.GetFileStreamByPackageVersion(
> 		ctx,
> 		pvs[0],
> 		&packages_service.PackageFileInfo{
> 			Filename: filename,
> 		},
> 	)
> 	if err != nil {
> 		if err == packages_model.ErrPackageFileNotExist {
> 			apiError(ctx, http.StatusNotFound, err)
> 			return
> 		}
> 		apiError(ctx, http.StatusInternalServerError, err)
> 		return
> 	}
> 	defer s.Close()
> 
> 	ctx.ServeContent(pf.Name, s, pf.CreatedUnix.AsLocalTime())
> }
> ```
> 
> Then there is no regex parsing necessary. Which fails btw for `pack-age-1.0.0-v1-beta.0.tgz`.

Definitely a better way to do it, changes updated except I kept the string empty check so to reduce code duplication.  Can keep them more separate though if desired.

**Comment by Racer159** [comment]
> 

And changed it back after looking more at the code (since it would double lookup packages).  Would it make more sense to just drop DownloadPackageFile now and just always call DownloadPackageFileByName since both have to lookup the version?

**Review by KN4CK3R — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-M work

**Comment by KN4CK3R** [comment]
Why is this a `skip-changelog`? I think that should be `enhancement` or `feature`.

**Comment by lunny** [comment]
> Why is this a `skip-changelog`? I think that should be `enhancement` or `feature`.

skip-changelog means it should not be displayed in v1.19's changelog. It's not conflict with `enhancement` or `feature`.

**Pull request merged**